clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java

```diff
@@ -2000,6 +2003,58 @@ protected void processCommonConsumerAction(
     }
   }
 
+  /**
+   * This function checks, for a regular user store, if its persisted offset is greater than the end offset of its
+   * corresponding Kafka topic (indicating an offset rewind event, thus a potential data loss in Kafka) and increases
+   * related sensor counter values.
+   */
+  private void reportUserStoreOffsetRewindMetrics(PartitionConsumptionState pcs) {
+    if (!isUserSystemStore()) {
+      return;
+    }
+
+    long offset = pcs.getLatestProcessedLocalVersionTopicOffset();
+    if (offset == OffsetRecord.LOWEST_OFFSET) {
+      return;
+    }
+    // Proceed if persisted OffsetRecord exists and has meaningful content.
+    long endOffset = getKafkaTopicPartitionEndOffSet(localKafkaServer, kafkaVersionTopic, pcs.getPartition());
+    if (endOffset != StatsErrorCode.LAG_MEASUREMENT_FAILURE.code && offset > endOffset) {
+      // report offset rewind.
+      LOGGER.warn(
+          "Offset rewind for version topic: {}, persisted record offset: {}, Kafka topic partition end-offset: {}",
```

Comment:
We should also log the partition number.

Reply:
Good point.